### PR TITLE
AdaptiveHistogramEqualization and RescaleIntensity

### PR DESCRIPTION
**Pull request by MohamedOmar2020**
Wrapper for equalize_adapthist and rescale_intensity functions from skimage.

### CODE REVIEW

File: pathml/preprocessing/transforms.py [hunk 1]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
         tile.image = self.F(tile.image)
 
 
+
+
+class RescaleIntensity(Transform):
+    """
+    Adaptive histogram equalization
+
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py:135]: Done

File: pathml/preprocessing/transforms.py [hunk 2]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+    Args:
+        in_range, out_range (str or 2-tuple, optional): Min and max intensity values of input and output image.
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: Done

File: pathml/preprocessing/transforms.py [hunk 3]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+    def F(self, image):
+        assert image.dtype == np.uint8, f"image dtype {image.dtype} must be np.uint8"
+        image =  rescale_intensity(image, in_range=self.in_range, out_range=self.out_range)
+        return image
+
+    def apply(self, tile):
+        assert isinstance(
+            tile, pathml.core.tile.Tile
+        ), f"tile is type {type(tile)} but must be pathml.core.tile.Tile"
+        tile.image = tile.image.astype(np.uint8)
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: Not necessarily. I changed this part

File: pathml/preprocessing/transforms.py [hunk 4]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
 
+
+
+class RescaleIntensity(Transform):
+    """
+    Adaptive histogram equalization
+
+    Args:
+        in_range, out_range (str or 2-tuple, optional): Min and max intensity values of input and output image.
+
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: okie dokie

File: pathml/preprocessing/transforms.py [hunk 5]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+class AdaptiveHistogramEqualization(Transform):
+    """
+    Adaptive histogram equalization
+
+    Args:
+        kernel_size (int or array_like, optional): Defines the shape of contextual regions used in the algorithm. If iterable is passed, it must have the same number of elements as image.ndim (without color channel). If integer, it is broadcasted to each image dimension. By default, kernel_size is 1/8 of image height by 1/8 of its width.
+        clip_limit (float): Clipping limit, normalized between 0 and 1 (higher values give more contrast).
+        nbins (int): Number of gray bins for histogram (“data range”).
+
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: done

File: pathml/preprocessing/transforms.py [hunk 6]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+            tile, pathml.core.tile.Tile
+        ), f"tile is type {type(tile)} but must be pathml.core.tile.Tile"
+        tile.image = tile.image.astype(np.uint8)
+        tile.image = self.F(tile.image)
+
+
+class AdaptiveHistogramEqualization(Transform):
+    """
+    Adaptive histogram equalization
+
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py:195]: done

File: pathml/preprocessing/transforms.py [hunk 7]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+        assert image.dtype == np.uint8, f"image dtype {image.dtype} must be np.uint8"
+        image =  equalize_adapthist(image, kernel_size=self.kernel_size, clip_limit=self.clip_limit, nbins=self.nbins)
+        image = img_as_ubyte(image)
+        return image
+
+    def apply(self, tile):
+        assert isinstance(
+            tile, pathml.core.tile.Tile
+        ), f"tile is type {type(tile)} but must be pathml.core.tile.Tile"
+        tile.image = tile.image.astype(np.uint8)
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: changed this part

File: pathml/preprocessing/transforms.py [hunk 8]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
...
+    def __init__(self, kernel_size=None, clip_limit = 0.3, nbins=256):
+        self.kernel_size = kernel_size
+        self.clip_limit = clip_limit
+        self.nbins = nbins
+
+    def __repr__(self):
+        return f"AdaptiveHistogramEqualization(kernel_size={self.kernel_size}, clip_limit={self.clip_limit}, nbins={self.nbins})"
+
+    def F(self, image):
+        assert image.dtype == np.uint8, f"image dtype {image.dtype} must be np.uint8"
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: changed this part

File: pathml/preprocessing/transforms.py [hunk 9]
```diff
@@ -127,6 +130,72 @@ def apply(self, tile):
         tile.image = self.F(tile.image)
 
 
+
+
+class RescaleIntensity(Transform):
+    """
+    Adaptive histogram equalization
```
[the remark this answers is not in the document]
Comment (MohamedOmar2020) [reply pathml/preprocessing/transforms.py]: done

**Comment by MohamedOmar2020** [comment]
I added another function for histogram equalization

**Comment by ryanccarelli** [comment]
Closing this PR since these changes are added in #167

**Pull request closed without merging**